### PR TITLE
Remove `Banned` connection state from peer database

### PR DESCRIPTION
**Pull request by jmcph4**
## Issue Addressed

#2772

## Proposed Changes

 - Add `banned_since` field to `PeerInfo` type, capturing the time at which a peer was banned
 - Add accessors for this field
   - `banned_since`
   - `banned_since_mut`
 - Remove references to the old `PeerConnectionStatus::Banned` state
 - Correct the definition of whether a peer is banned or not to rely solely on the `ScoreState::Banned` variant

## Additional Info

N/A

### CODE REVIEW

**Review by divagant-martian — CHANGES_REQUESTED** [review]
while this obviously compiles and indeed removes the `PeerConnectionState::Banned`, it's not fully untangling score and connection state, since `NewConnectionState::Banned` and `NewConnectionState::Unbanned` still exist. 

Tests are failing because of this in part. `NewConnectionState::Banned` is adding banned peers and their ips, and later checks are removing them

File: beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs [hunk 1]
```diff
@@ -22,6 +22,9 @@ use PeerConnectionStatus::*;
 pub struct PeerInfo<T: EthSpec> {
     /// The peers reputation
     score: Score,
+    /// Time at which peer was banned
+    #[serde(skip)]
+    banned_since: Option<Instant>,
```
Comment (divagant-martian) [inline beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs:27]: I'm not sure this is the best way to keep track of this. Banned peers exist outside the `PeerInfo`, this keeps the possibility of inconsistency between that tracking and the inner one

**Comment by jmcph4** [comment]
> while this obviously compiles and indeed removes the `PeerConnectionState::Banned`, it's not fully untangling score and connection state, since `NewConnectionState::Banned` and `NewConnectionState::Unbanned` still exist.
> 
> Tests are failing because of this in part. `NewConnectionState::Banned` is adding banned peers and their ips, and later checks are removing them

I was thinking about this over the weekend and I don't get why `NewConnectionState` is strictly necessary? If you model the current connection for any given peer as a finite state machine, it doesn't really make sense to make a distinction between current state and new state -- they should be the same. When you compare the two different types it seems to reinforce this belief too:

https://github.com/sigp/lighthouse/blob/0d6530a0163eca1a0f39e3bab4465c690e41dc7d/beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs#L483-L511

https://github.com/sigp/lighthouse/blob/0d6530a0163eca1a0f39e3bab4465c690e41dc7d/beacon_node/lighthouse_network/src/peer_manager/peerdb.rs#L1110-L1139

Eager for your insight -- perhaps I'm oversimplifying it!

**Comment by divagant-martian** [comment]
So this was added by age back in https://github.com/sigp/lighthouse/pull/2660.

It took me a bit to understand why it made sense again. The main difference is that the `ConnectionState` is pub while the `NewConnectionState` is private. This was made to isolate the connection state from outside modification since its container (the `PeerInfo`) needs to be mutated outside the `PeerManager` as well.  I would be on board with creating perhaps a newtype to encapulate this without duplicating the state, achieves the same and maintains the isolation. That is a bit outside the scope of this tho, as the focus is first to stop the entanglement between score and connection state

**Pull request closed without merging**